net/WebSocketHandler.hpp

```diff
@@ -241,21 +246,52 @@ class WebSocketHandler : public ProtocolHandlerInterface
     void shutdown(const StatusCodes statusCode = StatusCodes::NORMAL_CLOSE,
                   const std::string& statusMessage = std::string(),
                   bool hardShutdown = false)
+    {
+        shutdownImpl(statusCode, statusMessage, hardShutdown, false);
+    }
+
+    /// Don't wait for the remote Websocket to handshake with us; go down fast.
+    void shutdownAfterWriting()
+    {
+        shutdownImpl(WebSocketHandler::StatusCodes::NORMAL_CLOSE, std::string(),
+                     true /* hard async shutdown & close */, false);
+    }
+
+    /// Returns true if the underlying socket is connected.
+    bool isConnected() const {
+        std::shared_ptr<StreamSocket> socket = _socket.lock();
+        return nullptr != socket && !socket->isClosed();
+    }
+
+private:
+    void shutdownSilent() {
+        shutdownImpl(WebSocketHandler::StatusCodes::POLICY_VIOLATION /* ignored */,
+                     std::string(), true /* hard async shutdown & close */, true);
+    }
+    void shutdownImpl(const StatusCodes statusCode,
+                      const std::string& statusMessage,
+                      bool hardShutdown,
+                      bool silentShutdown)
     {
         std::shared_ptr<StreamSocket> socket = _socket.lock();
         if (socket)
         {
-            LOGA_TRC(WebSocket, "Shutdown: Closing Connection");
-            if (!_shuttingDown)
+            const bool silent = _shuttingDown || silentShutdown;
+            LOGA_TRC(WebSocket, "Shutdown: Closing Connection " << (silent ? "(silent)" : "(coop)"));
+            if (!silent)
+            {
                 sendCloseFrame(statusCode, statusMessage);
-            socket->closeConnection();
+            }
+            socket->shutdown();
```

Comment:
This is confusing. What does `closeConnection()` do, and what does `shutdown()` do? Do we really need this distinction?

Reply:
That was my 1st thought reading the Socket code as well.

However, this change fixes its usage and realigns with other cases,
i.e. overridden StreamSocket::shutdown() only does the async-signalling (intention)
while StreamSocket::closeConnection() does the Socket::Shutdown() ..
I surely don't like it, just played along. 

I am more than willing to propose a cleanup in an upcoming commit post this issue 9833